### PR TITLE
Make `Broadcast.flatten(bc).f` more complier frendly. (better inferred and inlined)

### LINKED ISSUE #27988: inference failure for broadcast implementations that use `flatten` (eg SparseArrays)
I've been using the new broadcasting API to implement broadcasting for some custom types (which btw, the new API is great and massively simplified my code, thanks!). I did however run into the following surprising inference failure which I've reduced down to the following code: 

```julia
import Base.Broadcast: BroadcastStyle, materialize, broadcastable
using Base.Broadcast: Broadcasted, Style, flatten
using Test

struct NumWrapper{T}
    data::T
end

broadcastable(n::NumWrapper) = n
BroadcastStyle(::Type{N}) where {N<:NumWrapper} = Style{N}()
broadcast_data(n::NumWrapper) = (n.data,)
function materialize(bc::Broadcasted{Style{N}}) where {N<:NumWrapper}
    flat_bc = flatten(bc)
    N(broadcast.(flat_bc.f, broadcast_data.(flat_bc.args)...)...)
end

foo(a,b) = @. a * a + (b * a) * b
bar(a,b) = @. a * a + b * a * b

@inferred foo(NumWrapper(1), NumWrapper(2)) #inferred as Any
@inferred bar(NumWrapper(1), NumWrapper(2)) #inferred correctly

```

As you can see, each `NumWrapper` just holds a number, and broadcasting over e.g. `a::NumWrapper .+ b::NumWrapper` becomes `NumWrapper(broadcast.(+,(a.data,),(b.data,))...) = NumWrapper((a.data .+ b.data,)...)`. Note I need the `.data` wrapped in a tuple in my real code because in the real case `NumWrapper` has multiple fields; this is also crucial to triggering the bug, although in the simple example above it probably seems unnecessary. In any case, you see that the seemingly unimportant addition of the parenthesis around `(b*a)` spoils type stability. This is on commit 656d58782c. 

Beyond a possible solution, I'm also curious if there's a workaround or a better way to code this up, I can't say I'm 100% sure I've used the new API as intended (in particular, I've used `flatten` and the docs make it seem like this shouldn't usually be necessary). In any case, hope this helps!

EDIT: Fixed a small mistake in the text describing what `a::NumWrapper .+ b::NumWrapper` became.

### LINKED ISSUE #47493: Performance problems when using broadcasting and logical operators
Broadcasting with boolean operators seems to be very slow in some circumstances:

See below one example with Julia 1.8.2:

```julia
julia> x = randn((100,));

julia> y = randn((1,100));

julia> @time (x.^2 .< 1) .&& (1 .+ y.^2  .<1);
  0.004369 seconds (144.73 k allocations: 3.005 MiB)

julia> @btime ($x.^2 .< 1) .&& (1 .+ $y.^2  .<1);
  4.009 ms (144713 allocations: 3.00 MiB)

# this works
julia> @btime ($x.^2 .< 1) .&& ($y.^2  .<1);
  2.523 μs (3 allocations: 5.56 KiB)

julia> @btime ($x.^2 .< 1) .|| ($y.^2 .+ 1 .<1);
  2.738 μs (3 allocations: 5.56 KiB)
```

### LINKED ISSUE #50554: Broadcasting flatten with type arguments causes type-instability
If `Base.Broadcast.flatten` is called on a `Broadcasted` object with a `DataType` `f`, the resulting `Broadcasted` object is no longer type-stable, triggering allocations.

This can be triggered when using `||` or `&&` in expressions, a simple reproducible example of this:
```julia
julia> x = rand(5); y = rand(5);

julia> f!(y, x) = @. y = ifelse(false || x < Float32(1), x, y)
f! (generic function with 1 method)

julia> f!(y, x);

julia> @allocated f!(y, x)
800
```

I suspect that this will fail altogether on the GPU.

### PR DESCRIPTION
**Pull request by N5N3**
A follow up attemp to fix #27988. (close #47493 close #50554)
Examples:
```julia
julia> using LazyArrays
julia> bc = @~ @. 1*(1 + 1) + 1*1;
julia> bc2 = @~ 1 .* 1 .- 1 .* 1 .^2 .+ 1 .* 1 .+ 1 .^ 3;
```
On master:
<details><summary> click for details </summary>
<p>

```julia
julia> @code_typed Broadcast.flatten(bc).f(1,1,1,1,1)
CodeInfo(
1 ─ %1  = Core.getfield(args, 1)::Int64
│   %2  = Core.getfield(args, 2)::Int64
│   %3  = Core.getfield(args, 3)::Int64
│   %4  = Core.getfield(args, 4)::Int64
│   %5  = Core.getfield(args, 5)::Int64
│   %6  = invoke Base.Broadcast.var"#13#14"{Base.Broadcast.var"#16#18"{Base.Broadcast.var"#15#17", Base.Broadcast.var"#13#14"{Base.Broadcast.var"#13#14"{Base.Broadcast.var"#15#17"}}, Base.Broadcast.var"#23#24"{Base.Broadcast.var"#23#24"{Base.Broadcast.var"#25#26"}}, Base.Broadcast.var"#19#20"{Base.Broadcast.var"#19#20"{Base.Broadcast.var"#21#22"}}, typeof(+)}}(Base.Broadcast.var"#16#18"{Base.Broadcast.var"#15#17", Base.Broadcast.var"#13#14"{Base.Broadcast.var"#13#14"{Base.Broadcast.var"#15#17"}}, Base.Broadcast.var"#23#24"{Base.Broadcast.var"#23#24"{Base.Broadcast.var"#25#26"}}, Base.Broadcast.var"#19#20"{Base.Broadcast.var"#19#20"{Base.Broadcast.var"#21#22"}}, typeof(+)}(Base.Broadcast.var"#15#17"(), Base.Broadcast.var"#13#14"{Base.Broadcast.var"#13#14"{Base.Broadcast.var"#15#17"}}(Base.Broadcast.var"#13#14"{Base.Broadcast.var"#15#17"}(Base.Broadcast.var"#15#17"())), Base.Broadcast.var"#23#24"{Base.Broadcast.var"#23#24"{Base.Broadcast.var"#25#26"}}(Base.Broadcast.var"#23#24"{Base.Broadcast.var"#25#26"}(Base.Broadcast.var"#25#26"())), Base.Broadcast.var"#19#20"{Base.Broadcast.var"#19#20"{Base.Broadcast.var"#21#22"}}(Base.Broadcast.var"#19#20"{Base.Broadcast.var"#21#22"}(Base.Broadcast.var"#21#22"())), +))(%1::Int64, %2::Int64, %3::Vararg{Int64}, %4, %5)::Tuple{Int64, Int64, Vararg{Int64}}
│   %7  = Core._apply_iterate(Base.iterate, Base.Broadcast.var"#19#20"{Base.Broadcast.var"#19#20"{Base.Broadcast.var"#21#22"}}(Base.Broadcast.var"#19#20"{Base.Broadcast.var"#21#22"}(Base.Broadcast.var"#21#22"())), %6)::Tuple{Int64, Int64}
│   %8  = Core._apply_iterate(Base.iterate, Base.Broadcast.var"#23#24"{Base.Broadcast.var"#23#24"{Base.Broadcast.var"#25#26"}}(Base.Broadcast.var"#23#24"{Base.Broadcast.var"#25#26"}(Base.Broadcast.var"#25#26"())), %6)::Tuple{Vararg{Int64}}
│   %9  = Core._apply_iterate(Base.iterate, Base.Broadcast.var"#16#18"{Base.Broadcast.var"#9#11", Base.Broadcast.var"#13#14"{Base.Broadcast.var"#13#14"{Base.Broadcast.var"#15#17"}}, Base.Broadcast.var"#23#24"{Base.Broadcast.var"#23#24"{Base.Broadcast.var"#25#26"}}, Base.Broadcast.var"#19#20"{Base.Broadcast.var"#19#20"{Base.Broadcast.var"#21#22"}}, typeof(*)}(Base.Broadcast.var"#9#11"(), Base.Broadcast.var"#13#14"{Base.Broadcast.var"#13#14"{Base.Broadcast.var"#15#17"}}(Base.Broadcast.var"#13#14"{Base.Broadcast.var"#15#17"}(Base.Broadcast.var"#15#17"())), Base.Broadcast.var"#23#24"{Base.Broadcast.var"#23#24"{Base.Broadcast.var"#25#26"}}(Base.Broadcast.var"#23#24"{Base.Broadcast.var"#25#26"}(Base.Broadcast.var"#25#26"())), Base.Broadcast.var"#19#20"{Base.Broadcast.var"#19#20"{Base.Broadcast.var"#21#22"}}(Base.Broadcast.var"#19#20"{Base.Broadcast.var"#21#22"}(Base.Broadcast.var"#21#22"())), *), %8)::Tuple{Int64}
│   %10 = Core.getfield(%7, 1)::Int64
│   %11 = Core.getfield(%7, 2)::Int64
│   %12 = Base.mul_int(%10, %11)::Int64
│   %13 = Core.getfield(%9, 1)::Int64
│   %14 = Base.add_int(%12, %13)::Int64
└──       return %14
) => Int64

julia> @code_typed Broadcast.flatten(bc2).f(1,1,1,^,1,Val(2),1,1,^,1,Val(3))
CodeInfo(
1 ─ %1  = Core.getfield(args, 1)::Int64
│   %2  = Core.getfield(args, 2)::Int64
│   %3  = Core.getfield(args, 3)::Int64
│   %4  = Core.getfield(args, 5)::Int64
│   %5  = Core.getfield(args, 7)::Int64
│   %6  = Core.getfield(args, 8)::Int64
│   %7  = Core.getfield(args, 10)::Int64
│   %8  = invoke Base.Broadcast.var"#13#14"{Base.Broadcast.var"#16#18"{Base.Broadcast.var"#15#17", Base.Broadcast.var"#13#14"{Base.Broadcast.var"#13#14"{Base.Broadcast.var"#13#14"{Base.Broadcast.var"#15#17"}}}, Base.Broadcast.var"#23#24"{Base.Broadcast.var"#23#24"{Base.Broadcast.var"#23#24"{Base.Broadcast.var"#25#26"}}}, Base.Broadcast.var"#19#20"{Base.Broadcast.var"#19#20"{Base.Broadcast.var"#19#20"{Base.Broadcast.var"#21#22"}}}, typeof(Base.literal_pow)}}(Base.Broadcast.var"#16#18"{Base.Broadcast.var"#15#17", Base.Broadcast.var"#13#14"{Base.Broadcast.var"#13#14"{Base.Broadcast.var"#13#14"{Base.Broadcast.var"#15#17"}}}, Base.Broadcast.var"#23#24"{Base.Broadcast.var"#23#24"{Base.Broadcast.var"#23#24"{Base.Broadcast.var"#25#26"}}}, Base.Broadcast.var"#19#20"{Base.Broadcast.var"#19#20"{Base.Broadcast.var"#19#20"{Base.Broadcast.var"#21#22"}}}, typeof(Base.literal_pow)}(Base.Broadcast.var"#15#17"(), Base.Broadcast.var"#13#14"{Base.Broadcast.var"#13#14"{Base.Broadcast.var"#13#14"{Base.Broadcast.var"#15#17"}}}(Base.Broadcast.var"#13#14"{Base.Broadcast.var"#13#14"{Base.Broadcast.var"#15#17"}}(Base.Broadcast.var"#13#14"{Base.Broadcast.var"#15#17"}(Base.Broadcast.var"#15#17"()))), Base.Broadcast.var"#23#24"{Base.Broadcast.var"#23#24"{Base.Broadcast.var"#23#24"{Base.Broadcast.var"#25#26"}}}(Base.Broadcast.var"#23#24"{Base.Broadcast.var"#23#24"{Base.Broadcast.var"#25#26"}}(Base.Broadcast.var"#23#24"{Base.Broadcast.var"#25#26"}(Base.Broadcast.var"#25#26"()))), Base.Broadcast.var"#19#20"{Base.Broadcast.var"#19#20"{Base.Broadcast.var"#19#20"{Base.Broadcast.var"#21#22"}}}(Base.Broadcast.var"#19#20"{Base.Broadcast.var"#19#20"{Base.Broadcast.var"#21#22"}}(Base.Broadcast.var"#19#20"{Base.Broadcast.var"#21#22"}(Base.Broadcast.var"#21#22"()))), Base.literal_pow))(%3::Int64, ^::Function, %4::Vararg{Any}, $(QuoteNode(Val{2}())), %5, %6, ^, %7, $(QuoteNode(Val{3}())))::Tuple{Int64, Any, Vararg{Any}}
│   %9  = Core._apply_iterate(Base.iterate, Base.Broadcast.var"#19#20"{Base.Broadcast.var"#19#20"{Base.Broadcast.var"#21#22"}}(Base.Broadcast.var"#19#20"{Base.Broadcast.var"#21#22"}(Base.Broadcast.var"#21#22"())), %8)::Tuple{Int64, Any}
│   %10 = Core._apply_iterate(Base.iterate, Base.Broadcast.var"#23#24"{Base.Broadcast.var"#23#24"{Base.Broadcast.var"#25#26"}}(Base.Broadcast.var"#23#24"{Base.Broadcast.var"#25#26"}(Base.Broadcast.var"#25#26"())), %8)::Tuple
│   %11 = Core._apply_iterate(Base.iterate, Base.Broadcast.var"#15#17"(), %10)::Tuple
│   %12 = Core.getfield(%9, 1)::Int64
│   %13 = Core.getfield(%9, 2)::Any
│   %14 = (*)(%12, %13)::Any
│   %15 = Core.tuple(%14)::Tuple{Any}
│   %16 = Core._apply_iterate(Base.iterate, Core.tuple, %15, %11)::Tuple{Any, Vararg{Any}}
│   %17 = Base.mul_int(%1, %2)::Int64
│   %18 = Core.tuple(%17)::Tuple{Int64}
│   %19 = Core._apply_iterate(Base.iterate, Core.tuple, %18, %16)::Tuple{Int64, Any, Vararg{Any}}
│   %20 = Core._apply_iterate(Base.iterate, Base.Broadcast.var"#19#20"{Base.Broadcast.var"#19#20"{Base.Broadcast.var"#21#22"}}(Base.Broadcast.var"#19#20"{Base.Broadcast.var"#21#22"}(Base.Broadcast.var"#21#22"())), %19)::Tuple{Int64, Any}
│   %21 = Core._apply_iterate(Base.iterate, Base.Broadcast.var"#23#24"{Base.Broadcast.var"#23#24"{Base.Broadcast.var"#25#26"}}(Base.Broadcast.var"#23#24"{Base.Broadcast.var"#25#26"}(Base.Broadcast.var"#25#26"())), %19)::Tuple
│   %22 = Core._apply_iterate(Base.iterate, Base.Broadcast.var"#16#18"{Base.Broadcast.var"#15#17", Base.Broadcast.var"#13#14"{Base.Broadcast.var"#13#14"{Base.Broadcast.var"#15#17"}}, Base.Broadcast.var"#23#24"{Base.Broadcast.var"#23#24"{Base.Broadcast.var"#25#26"}}, Base.Broadcast.var"#19#20"{Base.Broadcast.var"#19#20"{Base.Broadcast.var"#21#22"}}, typeof(*)}(Base.Broadcast.var"#15#17"(), Base.Broadcast.var"#13#14"{Base.Broadcast.var"#13#14"{Base.Broadcast.var"#15#17"}}(Base.Broadcast.var"#13#14"{Base.Broadcast.var"#15#17"}(Base.Broadcast.var"#15#17"())), Base.Broadcast.var"#23#24"{Base.Broadcast.var"#23#24"{Base.Broadcast.var"#25#26"}}(Base.Broadcast.var"#23#24"{Base.Broadcast.var"#25#26"}(Base.Broadcast.var"#25#26"())), Base.Broadcast.var"#19#20"{Base.Broadcast.var"#19#20"{Base.Broadcast.var"#21#22"}}(Base.Broadcast.var"#19#20"{Base.Broadcast.var"#21#22"}(Base.Broadcast.var"#21#22"())), *), %21)::Tuple{Any, Vararg{Any}}
│   %23 = Core.getfield(%20, 1)::Int64
│   %24 = Core.getfield(%20, 2)::Any
│   %25 = (-)(%23, %24)::Any
│   %26 = Core.tuple(%25)::Tuple{Any}
│   %27 = Core._apply_iterate(Base.iterate, Core.tuple, %26, %22)::Tuple{Any, Any, Vararg{Any}}
│   %28 = Core._apply_iterate(Base.iterate, Base.Broadcast.var"#19#20"{Base.Broadcast.var"#19#20"{Base.Broadcast.var"#21#22"}}(Base.Broadcast.var"#19#20"{Base.Broadcast.var"#21#22"}(Base.Broadcast.var"#21#22"())), %27)::Tuple{Any, Any}
│   %29 = Core._apply_iterate(Base.iterate, Base.Broadcast.var"#23#24"{Base.Broadcast.var"#23#24"{Base.Broadcast.var"#25#26"}}(Base.Broadcast.var"#23#24"{Base.Broadcast.var"#25#26"}(Base.Broadcast.var"#25#26"())), %27)::Tuple
│   %30 = Core._apply_iterate(Base.iterate, Base.Broadcast.var"#16#18"{Base.Broadcast.var"#9#11", Base.Broadcast.var"#13#14"{Base.Broadcast.var"#13#14"{Base.Broadcast.var"#13#14"{Base.Broadcast.var"#15#17"}}}, Base.Broadcast.var"#23#24"{Base.Broadcast.var"#23#24"{Base.Broadcast.var"#23#24"{Base.Broadcast.var"#25#26"}}}, Base.Broadcast.var"#19#20"{Base.Broadcast.var"#19#20"{Base.Broadcast.var"#19#20"{Base.Broadcast.var"#21#22"}}}, typeof(Base.literal_pow)}(Base.Broadcast.var"#9#11"(), Base.Broadcast.var"#13#14"{Base.Broadcast.var"#13#14"{Base.Broadcast.var"#13#14"{Base.Broadcast.var"#15#17"}}}(Base.Broadcast.var"#13#14"{Base.Broadcast.var"#13#14"{Base.Broadcast.var"#15#17"}}(Base.Broadcast.var"#13#14"{Base.Broadcast.var"#15#17"}(Base.Broadcast.var"#15#17"()))), Base.Broadcast.var"#23#24"{Base.Broadcast.var"#23#24"{Base.Broadcast.var"#23#24"{Base.Broadcast.var"#25#26"}}}(Base.Broadcast.var"#23#24"{Base.Broadcast.var"#23#24"{Base.Broadcast.var"#25#26"}}(Base.Broadcast.var"#23#24"{Base.Broadcast.var"#25#26"}(Base.Broadcast.var"#25#26"()))), Base.Broadcast.var"#19#20"{Base.Broadcast.var"#19#20"{Base.Broadcast.var"#19#20"{Base.Broadcast.var"#21#22"}}}(Base.Broadcast.var"#19#20"{Base.Broadcast.var"#19#20"{Base.Broadcast.var"#21#22"}}(Base.Broadcast.var"#19#20"{Base.Broadcast.var"#21#22"}(Base.Broadcast.var"#21#22"()))), Base.literal_pow), %29)::Tuple{Any}
│   %31 = Core.getfield(%28, 1)::Any
│   %32 = Core.getfield(%28, 2)::Any
│   %33 = (+)(%31, %32)::Any
│   %34 = Core.getfield(%30, 1)::Any
│   %35 = (+)(%33, %34)::Any
└──       return %35
) => Any
```
</p>

</details>

On this PR
```julia
julia> @code_typed Broadcast.flatten(bc).f(1,1,1,1,1)   
CodeInfo(
1 ─ %1 = Core.getfield(args, 1)::Int64
│   %2 = Core.getfield(args, 2)::Int64
│   %3 = Core.getfield(args, 3)::Int64
│   %4 = Core.getfield(args, 4)::Int64
│   %5 = Core.getfield(args, 5)::Int64
│   %6 = Base.add_int(%2, %3)::Int64
│   %7 = Base.mul_int(%1, %6)::Int64
│   %8 = Base.mul_int(%4, %5)::Int64
│   %9 = Base.add_int(%7, %8)::Int64
└──      return %9
) => Int64

julia> @code_typed Broadcast.flatten(bc2).f(1,1,1,^,1,Val(2),1,1,^,1,Val(3))
CodeInfo(
1 ─ %1  = Core.getfield(args, 1)::Int64
│   %2  = Core.getfield(args, 2)::Int64
│   %3  = Core.getfield(args, 3)::Int64
│   %4  = Core.getfield(args, 5)::Int64
│   %5  = Core.getfield(args, 7)::Int64
│   %6  = Core.getfield(args, 8)::Int64
│   %7  = Core.getfield(args, 10)::Int64
│   %8  = Base.mul_int(%1, %2)::Int64
│   %9  = Base.mul_int(%4, %4)::Int64
│   %10 = Base.mul_int(%3, %9)::Int64
│   %11 = Base.sub_int(%8, %10)::Int64
│   %12 = Base.mul_int(%5, %6)::Int64
│   %13 = Base.add_int(%11, %12)::Int64
│   %14 = Base.mul_int(%7, %7)::Int64
│   %15 = Base.mul_int(%14, %7)::Int64
│   %16 = Base.add_int(%13, %15)::Int64
└──       return %16
) => Int64
```

### CODE REVIEW

**Comment by N5N3** [comment]
Test for `BitArray`'s chunked `broadcast` passed locally after the above commit.
This PR does change the default behavior of `flatten` (add more type restriction on `bc.f`).
But I think this is not a break change, as `flatten` is used for `broadcast` mostly, and user could use `flatten(bc; replace = Returns(Any))` to recover the current behaviour if they dislike the type annotation.

**Comment by N5N3** [comment]
I only add the second example to test. As the first one's instability didnt influence the return type.
So the test might not be that robust. (Do we have tools to test internal instability?)
Ping previous contibutors to review. @mbauman @Keno

**Comment by N5N3** [comment]
Downloads error on Freebsd seems unrelated.

**Comment by N5N3** [comment]
I'm not sure why [52a3132](https://github.com/JuliaLang/julia/pull/43322/commits/52a3132d294e280952fe654ae78e9c80a8462020) helps our optimizer inline flattened `Broadcasted`. But it works. The code style is similar to https://github.com/JuliaLang/julia/pull/41139, but unnecessary extra methods is avoided.

Some Benchmark:

- Issue https://github.com/JuliaArrays/StaticArrays.jl/issues/560
```julia
ũ, u₀, u₁, ρ = rand(SVector{3}), rand(SVector{3}), rand(SVector{3}), rand(SVector{3})
f(ũ, u₀, u₁, ρ) = ũ ./ (1e-6 .+ max.(abs.(u₀), abs.(u₁)) .* ρ)
@time f(ũ, u₀, u₁, ρ)
#master:  0.396873 seconds (887.74 k allocations: 47.879 MiB, 4.19% gc time, 99.99% compilation time)
#dbb86b8: 0.453568 seconds (998.51 k allocations: 54.036 MiB, 5.23% gc time, 99.99% compilation time)
#52a3132: 0.535454 seconds (1.36 M allocations: 72.842 MiB, 5.36% gc time, 100.00% compilation time)
@btime f($ũ, $u₀, $u₁, $ρ)
#master:  79.609 ns (12 allocations: 192 bytes)
#dbb86b8: 85.862 ns (12 allocations: 192 bytes)
#52a3132: 6.300 ns (0 allocations: 0 bytes)
```
- Issue https://github.com/JuliaArrays/StaticArrays.jl/issues/682
```julia
_dual(x) = ForwardDiff.Dual(x, 1 + x)
f(η, B, R, 𝑢, σ, C) = (η .* (B .- R) ./ 𝑢 .- C) ./ σ
A = fill(1.0, SMatrix{2, 2})
b = 1.0
@time f(A, b, A, _dual(b), b, A)
#master:  0.654511 seconds (1.60 M allocations: 86.765 MiB, 6.85% gc time, 99.89% compilation time)
#dbb86b8: 0.687675 seconds (1.65 M allocations: 89.454 MiB, 7.66% gc time, 99.99% compilation time)
#52a3132: 0.703334 seconds (1.93 M allocations: 104.449 MiB, 7.47% gc time, 99.99% compilation time)
@btime f($A, $b, $A, $(_dual(b)), $b, $A)
#master:  12.000 μs (310 allocations: 7.34 KiB)
#dbb86b8: 14.314 ns (0 allocations: 0 bytes)
#52a3132: 14.314 ns (0 allocations: 0 bytes)
```
- Issue https://github.com/JuliaArrays/StaticArrays.jl/issues/609
```julia
doit(s, c::Integer) = (s .- s.^2) ./ c
s = @SVector[1.]
c = 2
@time doit(s,c)
#master:  0.428366 seconds (885.46 k allocations: 48.372 MiB, 3.97% gc time, 99.96% compilation time)
#dbb86b8: 0.430992 seconds (884.13 k allocations: 48.546 MiB, 5.82% gc time, 99.99% compilation time)
#52a3132: 0.484751 seconds (1.10 M allocations: 60.163 MiB, 4.64% gc time, 100.00% compilation time)
@btime doit($s,$c)
#master:  1.030 μs (24 allocations: 448 bytes)
#dbb86b8: 2.300 ns (0 allocations: 0 bytes)
#52a3132: 2.300 ns (0 allocations: 0 bytes)
```
- Issue https://github.com/JuliaArrays/StaticArrays.jl/issues/797
```julia
n = 5
a = MMatrix{n, n, Float64}(rand(n, n))
b = MMatrix{n, n, Float64}(rand(n, n))
c = MMatrix{n, n, Float64}(rand(n, n))
d = MMatrix{n, n, Float64}(rand(n, n))
r = MMatrix{n, n, Float64}(rand(n, n))
function g!(r,a,b,c,d)
    @. r = a + 5 * b + 3 * c - d
end
@time g!(r, a, b, c, d)
#master:  0.700729 seconds (2.41 M allocations: 105.241 MiB, 18.97% gc time, 99.97% compilation time)
#dbb86b8: 0.580067 seconds (2.79 M allocations: 123.188 MiB, 8.25% gc time, 100.00% compilation time)
#52a3132: 0.649758 seconds (3.11 M allocations: 140.541 MiB, 7.76% gc time, 100.00% compilation time)
@btime g!($r, $a, $b, $c, $d)
#master:  17.500 μs (750 allocations: 13.28 KiB)
#dbb86b8: 25.201 ns (0 allocations: 0 bytes)
#52a3132: 25.251 ns (0 allocations: 0 bytes)
```

**Comment by N5N3** [comment]
I'm not sure why 992d2d4 helps our optimizer inline flattened `Broadcasted`. But it works (although only one valid example). The code style is similar to https://github.com/JuliaLang/julia/pull/41139, but unnecessary extra methods are avoided.

Some Benchmark:

- Issue https://github.com/JuliaArrays/StaticArrays.jl/issues/560
```julia
ũ, u₀, u₁, ρ = rand(SVector{3}), rand(SVector{3}), rand(SVector{3}), rand(SVector{3})
f(ũ, u₀, u₁, ρ) = ũ ./ (1e-6 .+ max.(abs.(u₀), abs.(u₁)) .* ρ)
@time f(ũ, u₀, u₁, ρ)
#master:  0.396873 seconds (887.74 k allocations: 47.879 MiB, 4.19% gc time, 99.99% compilation time)
#dbb86b8: 0.453568 seconds (998.51 k allocations: 54.036 MiB, 5.23% gc time, 99.99% compilation time)
#992d2d4: 0.494349 seconds (1.05 M allocations: 56.624 MiB, 4.77% gc time, 99.99% compilation time)
@btime f($ũ, $u₀, $u₁, $ρ)
#master:  79.609 ns (12 allocations: 192 bytes)
#dbb86b8: 85.862 ns (12 allocations: 192 bytes)
#992d2d4: 6.300 ns (0 allocations: 0 bytes)
```
- Issue https://github.com/JuliaArrays/StaticArrays.jl/issues/682
```julia
_dual(x) = ForwardDiff.Dual(x, 1 + x)
f(η, B, R, 𝑢, σ, C) = (η .* (B .- R) ./ 𝑢 .- C) ./ σ
A = fill(1.0, SMatrix{2, 2})
b = 1.0
@time f(A, b, A, _dual(b), b, A)
#master:  0.654511 seconds (1.60 M allocations: 86.765 MiB, 6.85% gc time, 99.89% compilation time)
#dbb86b8: 0.687675 seconds (1.65 M allocations: 89.454 MiB, 7.66% gc time, 99.99% compilation time)
#992d2d4: 0.684861 seconds (1.70 M allocations: 92.613 MiB, 7.07% gc time, 99.99% compilation time)
@btime f($A, $b, $A, $(_dual(b)), $b, $A)
#master:  12.000 μs (310 allocations: 7.34 KiB)
#dbb86b8: 14.314 ns (0 allocations: 0 bytes)
#992d2d4: 14.314 ns (0 allocations: 0 bytes)
```
- Issue https://github.com/JuliaArrays/StaticArrays.jl/issues/609
```julia
doit(s, c::Integer) = (s .- s.^2) ./ c
s = @SVector[1.]
c = 2
@time doit(s,c)
#master:  0.428366 seconds (885.46 k allocations: 48.372 MiB, 3.97% gc time, 99.96% compilation time)
#dbb86b8: 0.430992 seconds (884.13 k allocations: 48.546 MiB, 5.82% gc time, 99.99% compilation time)
#992d2d4: 0.444262 seconds (873.48 k allocations: 47.796 MiB, 6.07% gc time, 99.98% compilation time)
@btime doit($s,$c)
#master:  1.030 μs (24 allocations: 448 bytes)
#dbb86b8: 2.300 ns (0 allocations: 0 bytes)
#992d2d4: 2.300 ns (0 allocations: 0 bytes)
```
- Issue https://github.com/JuliaArrays/StaticArrays.jl/issues/797
```julia
n = 5
a = MMatrix{n, n, Float64}(rand(n, n))
b = MMatrix{n, n, Float64}(rand(n, n))
c = MMatrix{n, n, Float64}(rand(n, n))
d = MMatrix{n, n, Float64}(rand(n, n))
r = MMatrix{n, n, Float64}(rand(n, n))
function g!(r,a,b,c,d)
    @. r = a + 5 * b + 3 * c - d
end
@time g!(r, a, b, c, d)
#master:  0.700729 seconds (2.41 M allocations: 105.241 MiB, 18.97% gc time, 99.97% compilation time)
#dbb86b8: 0.580067 seconds (2.79 M allocations: 123.188 MiB, 8.25% gc time, 100.00% compilation time)
#992d2d4: 0.639049 seconds (2.80 M allocations: 124.040 MiB, 11.91% gc time, 99.99% compilation time)
@btime g!($r, $a, $b, $c, $d)
#master:  17.500 μs (750 allocations: 13.28 KiB)
#dbb86b8: 25.201 ns (0 allocations: 0 bytes)
#992d2d4: 25.251 ns (0 allocations: 0 bytes)
```

**Comment by N5N3** [comment]
It turns out this problem could be overcomed much more easily, as nested `Broadcasted`'s `getindex` is well inferred and inlined by our compiler.
In [a8283f2], the process of "Broadcast.flatten" follows `Broadcast.preprocess`'s style: `make_makeargs` will generate a tuple of functions. These functions take in the whole "flatttened" argument list, and return the appropriate input arguments of `bc.f`. i.e.
```julia
flattened(args...) = bc.f(map(f -> f(args), make_makeargs)...)
```

**Comment by ChrisRackauckas** [comment]
Maybe @vchuravy can take a look at this? In general packages shouldn't use as much Broadcast.flatten as they do, but I think this will have a lot of knock-on effects because they do use it a lot, and the current lowering is pretty bad.

**Review by vtjnash — COMMENTED** [review]
This reminds me of #45789–was that the inspiration?

ah, this is a fun new way to make use of our compiler's constant prop. I feel a bit nervous that the compiler won't be up to the challenge though of figuring out the Pick{N} allocation, or that it should be a Val{N} instead (with the caller doing the indexing explicitly with the result rather than calling the result as a function). How well is that working though in your experience with this?

**Comment by N5N3** [comment]
In fact this was inspired by the current broadcast implement itself.
I just noticed that our nested `Broadcasted`'s `getindex`/`preprocess` could be well inferred and inlined even with indirect recursion.

As for `Pick{N}`, I haven't encounter any realistic problem with it.
I tested it with with more than 32 args, it turns out we touch the inline limitation of `tail` (in `Broadcast._getindex`) first, while `flatten` and "flattened function" still work well.
Thus my local version even tries to store `singleton instance`/`type`/`bit scalar` inputs into `Returns` to reduce the number of broadcasted arguments.

Since the flatten process is stable. I think our complier should be able to handle it well.
BTW is there a difference between `Pick{N}` and `Val{N}`?

**Comment by vtjnash** [comment]
No difference content or concept wise. Val just means you have to unwrap it yourself later and call getindex, rather than call it directly with the Tuple and have that defined to return the indexed value

File: base/broadcast.jl [hunk 1]
```diff
@@ -341,20 +341,16 @@ function flatten(bc::Broadcasted)
...
+    # build a tuple of functions `makeargs`. Its elements take
+    # the whole "flat" argument list and and generate the appropriate
+    # input arguments for the broadcasted function `f`, e.g.,
+    #          makeargs[1] = ((w, x, y, z)) -> w
+    #          makeargs[2] = ((w, x, y, z)) -> g(x, y)
+    #          makeargs[3] = ((w, x, y, z)) -> z
+    makeargs = make_makeargs(bc.args)
+    f = Base.maybeconstructor(bc.f)
+    newf = (args...) -> (@inline; f(prepare_args(makeargs, args)...))
+    return Broadcasted{Style}(newf, args, bc.axes)
```
Comment (vtjnash) [inline base/broadcast.jl:353]: bc.style should be passed explicitly now to this constructor

**Review by vtjnash — APPROVED** [review]

**Pull request merged**